File: plugin/openapi/src/main/java/io/gatehill/imposter/plugin/openapi/service/SchemaServiceImpl.java
package io.gatehill.imposter.plugin.openapi.service;

import io.gatehill.imposter.plugin.openapi.model.ContentTypedHolder;
import io.gatehill.imposter.plugin.openapi.util.RefUtil;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.media.ArraySchema;
import io.swagger.v3.oas.models.media.ComposedSchema;
import io.swagger.v3.oas.models.media.DateSchema;
import io.swagger.v3.oas.models.media.DateTimeSchema;
import io.swagger.v3.oas.models.media.ObjectSchema;
import io.swagger.v3.oas.models.media.Schema;
import io.vertx.core.http.HttpServerRequest;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import static java.util.Collections.emptyMap;
import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;

/**
 * Collects examples from schema definitions.
 *
 * @author benjvoigt
 * @author Pete Cornish {@literal <outofcoffee@gmail.com>}
 */
public class SchemaServiceImpl implements SchemaService {
    private static final Logger LOGGER = LogManager.getLogger(SchemaServiceImpl.class);
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE.withZone(ZoneId.from(ZoneOffset.UTC));
    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneId.from(ZoneOffset.UTC));

    private static final Map<String, DefaultValueProvider<?>> DEFAULT_VALUE_PROVIDERS = new HashMap<String, DefaultValueProvider<?>>() {{
        put("string", new StringDefaultValueProvider());
        put("number", new NumberDefaultValueProvider());
        put("integer", new IntegerDefaultValueProvider());
        put("boolean", new BooleanDefaultValueProvider());
    }};

    @Override
    public ContentTypedHolder<?> collectExamples(HttpServerRequest request, OpenAPI spec, ContentTypedHolder<Schema<?>> schema) {
        final Object example = collectSchemaExample(spec, schema.getValue());
        LOGGER.trace("Collected example from {} schema for {} {}: {}", schema.getContentType(), request.method(), request.absoluteURI(), example);

        return new ContentTypedHolder<>(schema.getContentType(), example);
    }

    private Object collectSchemaExample(OpenAPI spec, Schema<?> schema) {
        final Object example;

        // $ref takes precedence, per spec:
        //   "Any sibling elements of a $ref are ignored. This is because
        //   $ref works by replacing itself and everything on its level
        //   with the definition it is pointing at."
        // See: https://swagger.io/docs/specification/using-ref/
        if (nonNull(schema.get$ref())) {
            final Schema<?> referent = RefUtil.lookupSchemaRef(spec, schema);
            example = collectSchemaExample(spec, referent);

        } else if (nonNull(schema.getExample())) {
            if (schema instanceof DateTimeSchema) {
                example = DATE_TIME_FORMATTER.format((OffsetDateTime) schema.getExample());
            } else if (schema instanceof DateSchema) {
                example = DATE_FORMATTER.format(((Date) schema.getExample()).toInstant());
            } else {
                example = schema.getExample();
            }

        } else if (nonNull(schema.getProperties())) {
            example = buildFromProperties(spec, schema.getProperties());

        } else if (ObjectSchema.class.isAssignableFrom(schema.getClass())) {
            final ObjectSchema objectSchema = (ObjectSchema) schema;
            example = buildFromProperties(spec, objectSchema.getProperties());

        } else if (ArraySchema.class.isAssignableFrom(schema.getClass())) {
            example = buildFromArraySchema(spec, (ArraySchema) schema);

        } else if (ComposedSchema.class.isAssignableFrom(schema.getClass())) {
            example = buildFromComposedSchema(spec, (ComposedSchema) schema);

        } else {
            example = getPropertyDefault(schema);
        }

        return example;
    }

    private List<Object> buildFromArraySchema(OpenAPI spec, ArraySchema schema) {
        // items may be a schema type with multiple children
        final Schema<?> items = schema.getItems();
        final List<Object> examples = new ArrayList<>();
        examples.add(collectSchemaExample(spec, items));
        return examples;
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    private Object buildFromComposedSchema(OpenAPI spec, ComposedSchema schema) {
        final Object example;
        if (nonNull(schema.getAllOf()) && !schema.getAllOf().isEmpty()) {
            final List<Schema> allOf = schema.getAllOf();

            // Combine properties of 'allOf'
            // See: https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/
            final Map<String, Object> combinedExampleProperties = new HashMap<>();
            allOf.forEach(s -> {
                final Object exampleMap = collectSchemaExample(spec, s);
                if (nonNull(exampleMap) && exampleMap instanceof Map) {
                    // FIXME code defensively around this cast
                    combinedExampleProperties.putAll((Map<String, Object>) exampleMap);
                }
            });
            example = combinedExampleProperties;

        } else if (nonNull(schema.getOneOf()) && !schema.getOneOf().isEmpty()) {
            LOGGER.debug("Found 'oneOf' in schema {} - using first schema example", schema.getName());
            final List<Schema> oneOf = schema.getOneOf();
            example = collectSchemaExample(spec, oneOf.get(0));

        } else if (nonNull(schema.getAnyOf()) && !schema.getAnyOf().isEmpty()) {
            LOGGER.debug("Found 'anyOf' in schema {} - using first schema example", schema.getName());
            final List<Schema> anyOf = schema.getAnyOf();
            example = collectSchemaExample(spec, anyOf.get(0));

        } else if (nonNull(schema.getNot())) {
            LOGGER.debug("Found 'not' in schema {} - using null for schema example", schema.getName());
            example = null;

        } else {
            throw new IllegalStateException("Invalid composed schema; missing or empty [allOf, oneOf, anyOf]");
        }
        return example;
    }

    @SuppressWarnings("rawtypes")
    private Map<String, Object> buildFromProperties(OpenAPI spec, Map<String, Schema> properties) {
        if (isNull(properties)) {
            return emptyMap();
        }
        return properties.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> collectSchemaExample(spec, e.getValue())));
    }

    private Object getPropertyDefault(Schema<?> schema) {
        // if a non-empty enum exists, choose the first value
        if (nonNull(schema.getEnum()) && !schema.getEnum().isEmpty()) {
            return schema.getEnum().get(0);
        }

        // fall back to a default for the type
        if (nonNull(schema.getType())) {
            final DefaultValueProvider<?> defaultValueProvider = DEFAULT_VALUE_PROVIDERS.get(schema.getType());
            if (nonNull(defaultValueProvider)) {
                return defaultValueProvider.provide(schema);
            } else {
                LOGGER.warn("Unknown type: {} for schema: {} - returning null for example property", schema.getType(), schema.getName());
                return null;
            }
        }

        LOGGER.warn("Missing type for schema: {} - returning null for example property", schema.getName());
        return null;
    }

    private interface DefaultValueProvider<T> {
        T provide(Schema<?> schema);
    }

    private static class StringDefaultValueProvider implements DefaultValueProvider<String> {
        @Override
        public String provide(Schema<?> schema) {
            // TODO make these configurable
            if (nonNull(schema.getFormat())) {
                // see https://swagger.io/docs/specification/data-models/data-types/
                switch (schema.getFormat()) {
                    case "date":
                        return DATE_FORMATTER.format(LocalDateTime.now().truncatedTo(ChronoUnit.SECONDS));
                    case "date-time":
                        return DATE_TIME_FORMATTER.format(OffsetDateTime.now().truncatedTo(ChronoUnit.SECONDS));
                    case "password":
                        return "changeme";
                    case "byte":
                        // base64-encoded characters
                        return "SW1wb3N0ZXI0bGlmZQo=";
                    case "email":
                        return "test@example.com";
                    case "uuid":
                    case "guid":
                        return UUID.randomUUID().toString();
                }
            }
            return "";
        }
    }

    private static class NumberDefaultValueProvider implements DefaultValueProvider<Double> {
        @Override
        public Double provide(Schema<?> schema) {
            return 0.0;
        }
    }

    private static class IntegerDefaultValueProvider implements DefaultValueProvider<Integer> {
        @Override
        public Integer provide(Schema<?> schema) {
            return 0;
        }
    }

    private static class BooleanDefaultValueProvider implements DefaultValueProvider<Boolean> {
        @Override
        public Boolean provide(Schema<?> schema) {
            return false;
        }
    }
}

File: plugin/openapi/src/main/java/io/gatehill/imposter/plugin/openapi/service/SchemaServiceImpl.kt
/*
 * Copyright (c) 2016-2021.
 *
 * This file is part of Imposter.
 *
 * "Commons Clause" License Condition v1.0
 *
 * The Software is provided to you by the Licensor under the License, as
 * defined below, subject to the following condition.
 *
 * Without limiting other conditions in the License, the grant of rights
 * under the License will not include, and the License does not grant to
 * you, the right to Sell the Software.
 *
 * For purposes of the foregoing, "Sell" means practicing any or all of
 * the rights granted to you under the License to provide to third parties,
 * for a fee or other consideration (including without limitation fees for
 * hosting or consulting/support services related to the Software), a
 * product or service whose value derives, entirely or substantially, from
 * the functionality of the Software. Any license notice or attribution
 * required by the License must also include this Commons Clause License
 * Condition notice.
 *
 * Software: Imposter
 *
 * License: GNU Lesser General Public License version 3
 *
 * Licensor: Peter Cornish
 *
 * Imposter is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Imposter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Imposter.  If not, see <https://www.gnu.org/licenses/>.
 */
package io.gatehill.imposter.plugin.openapi.service

import io.gatehill.imposter.plugin.openapi.model.ContentTypedHolder
import io.gatehill.imposter.plugin.openapi.service.valueprovider.DEFAULT_VALUE_PROVIDERS
import io.gatehill.imposter.plugin.openapi.util.RefUtil
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.media.ArraySchema
import io.swagger.v3.oas.models.media.ComposedSchema
import io.swagger.v3.oas.models.media.DateSchema
import io.swagger.v3.oas.models.media.DateTimeSchema
import io.swagger.v3.oas.models.media.ObjectSchema
import io.swagger.v3.oas.models.media.Schema
import io.vertx.core.http.HttpServerRequest
import org.apache.logging.log4j.LogManager
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.*

/**
 * Collects examples from schema definitions.
 *
 * @author benjvoigt
 * @author Pete Cornish
 */
class SchemaServiceImpl : SchemaService {
    override fun collectExamples(
        request: HttpServerRequest,
        spec: OpenAPI,
        schema: ContentTypedHolder<Schema<*>>
    ): ContentTypedHolder<*> {
        val example = collectSchemaExample(spec, schema.value)
        LOGGER.trace(
            "Collected example from {} schema for {} {}: {}",
            schema.contentType,
            request.method(),
            request.absoluteURI(),
            example
        )
        return ContentTypedHolder(schema.contentType, example)
    }

    private fun collectSchemaExample(spec: OpenAPI, schema: Schema<*>): Any? {
        // $ref takes precedence, per spec:
        //   "Any sibling elements of a $ref are ignored. This is because
        //   $ref works by replacing itself and everything on its level
        //   with the definition it is pointing at."
        // See: https://swagger.io/docs/specification/using-ref/
        val example: Any? = if (Objects.nonNull(schema.`$ref`)) {
            val referent = RefUtil.lookupSchemaRef(spec, schema)
            collectSchemaExample(spec, referent)
        } else if (Objects.nonNull(schema.example)) {
            if (schema is DateTimeSchema) {
                DATE_TIME_FORMATTER.format(schema.getExample() as OffsetDateTime)
            } else if (schema is DateSchema) {
                DATE_FORMATTER.format((schema.getExample() as Date).toInstant())
            } else {
                schema.example
            }
        } else if (Objects.nonNull(schema.properties)) {
            buildFromProperties(spec, schema.properties)
        } else if (ObjectSchema::class.java.isAssignableFrom(schema.javaClass)) {
            val objectSchema = schema as ObjectSchema?
            buildFromProperties(spec, objectSchema!!.properties)
        } else if (ArraySchema::class.java.isAssignableFrom(schema.javaClass)) {
            buildFromArraySchema(spec, schema as ArraySchema?)
        } else if (ComposedSchema::class.java.isAssignableFrom(schema.javaClass)) {
            buildFromComposedSchema(spec, schema as ComposedSchema?)
        } else {
            getPropertyDefault(schema)
        }
        return example
    }

    private fun buildFromArraySchema(spec: OpenAPI, schema: ArraySchema?): List<Any?> {
        // items may be a schema type with multiple children
        val items = schema!!.items
        val examples: MutableList<Any?> = mutableListOf()
        examples.add(collectSchemaExample(spec, items))
        return examples
    }

    private fun buildFromComposedSchema(spec: OpenAPI, schema: ComposedSchema?): Any? {
        val example: Any? = if (Objects.nonNull(schema!!.allOf) && schema.allOf.isNotEmpty()) {
            val allOf = schema.allOf

            // Combine properties of 'allOf'
            // See: https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/
            val combinedExampleProperties: MutableMap<String, Any> = mutableMapOf()
            allOf.forEach { s ->
                val exampleMap = collectSchemaExample(spec, s)
                if (Objects.nonNull(exampleMap) && exampleMap is Map<*, *>) {
                    // FIXME code defensively around this cast
                    @Suppress("UNCHECKED_CAST")
                    combinedExampleProperties.putAll((exampleMap as Map<String, Any>?)!!)
                }
            }
            combinedExampleProperties

        } else if (Objects.nonNull(schema.oneOf) && schema.oneOf.isNotEmpty()) {
            LOGGER.debug(
                "Found 'oneOf' in schema {} - using first schema example", (schema.name ?: "")
            )
            val oneOf = schema.oneOf
            collectSchemaExample(spec, oneOf[0])

        } else if (Objects.nonNull(schema.anyOf) && schema.anyOf.isNotEmpty()) {
            LOGGER.debug(
                "Found 'anyOf' in schema {} - using first schema example", (schema.name ?: "")
            )
            val anyOf = schema.anyOf
            collectSchemaExample(spec, anyOf[0])

        } else if (Objects.nonNull(schema.not)) {
            LOGGER.debug(
                "Found 'not' in schema {} - using null for schema example", (schema.name ?: "")
            )
            null

        } else {
            throw IllegalStateException("Invalid composed schema; missing or empty [allOf, oneOf, anyOf]")
        }
        return example
    }

    private fun buildFromProperties(spec: OpenAPI, properties: Map<String, Schema<*>>): Map<String, Any?> {
        return if (Objects.isNull(properties)) {
            emptyMap<String, Any>()
        } else {
            properties.entries.associate { (k, v) -> k to collectSchemaExample(spec, v) }
        }
    }

    private fun getPropertyDefault(schema: Schema<*>?): Any? {
        // if a non-empty enum exists, choose the first value
        if (Objects.nonNull(schema!!.enum) && !schema.enum.isEmpty()) {
            return schema.enum[0]
        }

        // fall back to a default for the type
        if (Objects.nonNull(schema.type)) {
            val defaultValueProvider = DEFAULT_VALUE_PROVIDERS[schema.type]!!
            return if (Objects.nonNull(defaultValueProvider)) {
                defaultValueProvider.provide(schema)
            } else {
                LOGGER.warn(
                    "Unknown type: {} for schema: {} - returning null for example property",
                    schema.type,
                    schema.name
                )
                null
            }
        }
        LOGGER.warn("Missing type for schema: {} - returning null for example property", schema.name)
        return null
    }

    companion object {
        private val LOGGER = LogManager.getLogger(
            SchemaServiceImpl::class.java
        )
        val DATE_FORMATTER: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE.withZone(ZoneId.from(ZoneOffset.UTC))
        val DATE_TIME_FORMATTER: DateTimeFormatter =
            DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneId.from(ZoneOffset.UTC))
    }
}